workflow/realisations.py

```diff
@@ -0,0 +1,713 @@
+    """Realisation JSON parse error."""
+
+    pass
+
+
+@dataclasses.dataclass
+class RealisationConfiguration(ABC):
+    """Abstract base class for RealisationConfiguration."""
+
+    _config_key: ClassVar[str]
+    """The configuration key to save and load from in the realisation."""
+    _schema: ClassVar[Schema]
+    """The reference schema to validate against when reading from a realisation."""
+
+    def to_dict(self) -> dict[str, Any]:
+        """
+        Convert the object to a dictionary representation.
+
+        Returns
+        -------
+        dict
+            Dictionary representation of the object.
+        """
+        return dataclasses.asdict(self)
+
+    @classmethod
+    def read_from_realisation(cls, realisation_ffp: Path) -> Self:
+        """Read configuration from a realisation file.
+
+        Parameters
+        ----------
+        realisation_ffp : Path
+            The filepath to read from.
+
+        Returns
+        -------
+        RealisationConfiguration
+            The configuration loaded from the realisation filepath. The
+            configuration schema is looked up from `cls._config_key`
+            and the key within the config is specified
+            `cls._schema`.
+
+        Raises
+        ------
+        RealisationParseError
+            If the key in `cls._config_key` is not present in
+            the realisation filepath.
+        """
+        with open(realisation_ffp, "r", encoding="utf-8") as realisation_file_handle:
+            realisation_config = json.load(realisation_file_handle)
+            if cls._config_key not in realisation_config:
+                raise RealisationParseError(
+                    f"No {cls._config_key} in realisation configuration"
+                )
+        return cls(**cls._schema.validate(realisation_config[cls._config_key]))
+
+    @classmethod
+    def read_from_defaults(cls, defaults_version: DefaultsVersion) -> Self:
+        """Read default values for this configuration.
+
+        Parameters
+        ----------
+        defaults_version : DefaultsVersion
+            The default parameter version to load with.
+
+        Returns
+        -------
+        RealisationConfiguration
+            The configuration loaded from the defaults. The configuration
+            schema is looked up from `cls._config_key` and the key within
+            the config is specified `cls._schema`.
+
+        Raises
+        ------
+        RealisationParseError
+            If the key in `cls._config_key` is not present in the scientific
+            defaults configuration.
+        """
+        default_config = defaults.load_defaults(defaults_version)
+        if cls._config_key not in default_config:
+            raise RealisationParseError(
+                f"No {cls._config_key} in defaults configuration"
+            )
+        return cls(**cls._schema.validate(default_config[cls._config_key]))
+
+    @classmethod
+    def read_from_realisation_or_defaults(
```

Comment:
Would we want to support cases where the realisation.json has like 1 variable defined but then the rest are grabbed from the defaults?

Reply:
Possibly? I haven't had the need for this yet so I haven't added it. Cesar also hasn't had a need for this just get. I'll wait and see if this becomes relevant and then add it.

Reply:
Fair call, probably will happen at some point but guess we will cross that bridge when we get there